### ISSUE TITLE
ansilexer custom lexer/formatter colorizes pytest output poorly

### ISSUE THREAD

**Issue by blairconrad** [mosra/m.css#148]
Love the theme/plugin/css/ etc. Thanks.

However, I was trying to include some pytest output in a blog post and found the output to be badly colorized.

Text as rendered at the console by pytest:
![image](https://user-images.githubusercontent.com/3275797/82438863-b3f97300-9a67-11ea-8dc0-1ed3246e7ff5.png)

and as rendered in the blog post by `include`:
![image](https://user-images.githubusercontent.com/3275797/82439555-df309200-9a68-11ea-8c13-a32f52a5bde2.png)

output from  `pygmentize -v -x -l .\m.css\plugins\ansilexer.py:AnsiLexer -f .\m.css\plugins\ansilexer.py:HtmlAnsiFormatter .\pytest.txt`:

```
============================= test session starts =============================
platform win32 -- Python 3.8.1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: D:\Sandbox\LibraryHippo
plugins: requests-mock-1.8.0
collected 2 items

tests\unit\test_wpl.py .<span class="g g-AnsiGreen">.
         [100%]</span>

============================== 2 passed<span class="g g-AnsiGreen"> in 1.46s ==============================</span>
```
As you can see, there's no directive to colorize
* the first line
* the first `.` on the test\unit\test_wpl.py line
* the first half of the summary line

(I'm actually quite surprised that the first `.` on the test\unit\test_wpl.py line appears green in the output!)

The fault appears to be in the lexer, as when I run
`pygmentize -x -l .\m.css\plugins\ansilexer.py:AnsiLexer -f raw  pytest.txt`, I see

```code
Token.Text	'============================= test session starts ============================='
Token.Text	'\nplatform win32 -- Python 3.8.1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1\nrootdir: D:\\Sandbox\\LibraryHippo\nplugins: requests-mock-1.8.0\ncollected 2 items\n\ntests\\unit\\test_wpl.py '
Token.Text	'.'
Token.Generic.AnsiGreen	'.'
Token.Generic.AnsiGreen	'                                                [100%]'
Token.Text	'\n\n'
Token.Text	'============================== '
Token.Text	''
Token.Text	'2 passed'
Token.Generic.AnsiGreen	' in 1.46s'
Token.Generic.AnsiGreen	' =============================='
Token.Text	'\n'
```

[pytest.txt](https://github.com/mosra/m.css/files/4656422/pytest.txt)

I'm happy to investigate, although I've had problems running the m.css tests, possibly due to missing prereqs (and I'm running on Windows).

**Comment by blairconrad** [comment]
As I see it, the problem is that the lexer assumes that the SGR codes will come in pairs, almost. Well, this regex

```
'\x1b\\[(\\d+)(?:;(\\d+))?m([^\x1b]*)'
```

allows for one SGR, but if there's only one, the first match is taken to be the brightness, which it might not be. And `callback` becomes unhappy when the color is missing, falling back to emitting a `Text` token.

Since the SGR codes, except for "Reset / Normal" (`\x1b[0`) are somewhat additive, in that they build on the previous state, perhaps changing only colour or brightness (or appearing in pairs to do both), I think a comprehensive solution would need to retain the current brightness and colour values. I'd explore such a change if you're interested. 

An alternative implementation, which I think is inferior, but I thought of first, would be to just allow for missing brightness or colour codes, and for the colour setting to come before brightness, but still limit ourselves to up to 2 SGRs. 
It runs the risk that some input will contain a brightness-enabling-only SGR and the lexer will change the colour to default from whatever it was, but without that, a brightness-disabling-only SGR (`^[[0m`) would fail to reset the colour to default.

I have hacked up `ansilexer.py` with this latter change, and my input sample gives

![image](https://user-images.githubusercontent.com/3275797/82450306-8669f500-9a7a-11ea-9278-598a668da687.png)

Whose output I'm content with. It's (to my mind) an improvement over the original. Here's the change.

```diff
diff --git a/plugins/ansilexer.py b/plugins/ansilexer.py
index 945c519..66b9d48 100644
--- a/plugins/ansilexer.py
+++ b/plugins/ansilexer.py
@@ -32,10 +32,20 @@ class AnsiLexer(RegexLexer):
     name = 'Ansi escape lexer'
 
     def callback(lexer, match):
-        bright = match.group(1)
-        color = match.group(2)
+        code1 = match.group(1)
+        code2 = match.group(2)
         text = match.group(3)
 
+        bright = '0'
+        if code2 in ['0', '1']:
+            bright = code2
+            color = code1
+        elif code1 in ['0', '1']:
+            bright = code1
+            color = code2
+        else:
+            color = code1
+
         if int(bright) == 0:
             token = 'Generic.Ansi'
         elif int(bright) == 1:
@@ -60,7 +70,7 @@ class AnsiLexer(RegexLexer):
             token += 'Cyan'
         elif color == '37':
             token += 'White'
-        elif color == '39':
+        elif color == '39' or color is None:
             token += 'Default'
         else:
             yield (match.start(), Text, text)
```

if you're happy with this, I can clean things up, attempt a test, and submit a PR. Or if you prefer a state-monitoring solution that will probably work better in some edge cases, I can try that too!

**Comment by mosra** [comment]
Hi and thank you! Quite a lot people started using this theme lately, makes me happy :)

The ANSI color code parser is a total mess, sorry about that -- I wrote it on an as-needed basis to work "good enough" with the output of [Corrade::TestSuite](https://doc.magnum.graphics/corrade/classCorrade_1_1TestSuite_1_1Tester.html) and then a bunch of other things, but it's not done properly at all.

A PR would be great. If you can try the state-monitoring solution, that would be the best of course, but if you get stuck or it turns out to be a multi-week project, I'm fine with accepting just the minimal set of changes that makes it work for pytest. The lexer is tested [inside the m.code plugin test](https://github.com/mosra/m.css/tree/master/plugins/m/test/code), so please extend that to cover the new cases. You can run the plugin tests alone with 

```py
cd plugins
python -m unittest m/test/test_code.py
```

This should only require pelican and pygments (running all tests needs latex, matplotlib, graphviz, qrcode and a ton of other stuff which is a pain to install, especially on windows). Thanks in advance :)

**Comment by blairconrad** [comment]
> Quite a lot people started using this theme lately, makes me happy

Oh, good. It's very attractive, and the modularity and documentation are without peer, in my experience. I'm quite happy. Of course I did tinker a little for my own blog, but so far I'm keeping quite close to your original.

----
Thanks. I'll try the more ambitious change. Also, thanks for the note about the test. I'll pursue, but initial attempt to run still fails.

The comparison fails around line 42, where what appear to be Han characters are rendered differently in `output/page.html`. For example, following `#a5c9ea">&zwnj;`, the input has  bytes `e2 96 88 e2 96 88`, which I think are supposed to be 蠷蠷 (two instances of [U+8837](https://www.compart.com/en/unicode/U+8837)). In the rendered output, I see `c3 a2 e2 80 93 cb 86 c3 a2  e2 80 93 cb 86`, which is, I was going to say, the equivalent of taking the original bytes, interpreting them as individual ISO-8859-something codepoints, and translating to utf-8, but I'm not quite getting the conversions to work out.

For what it's worth, my system encoding is utf-8, so that might have something to do with it:
```powershell
❯ python -c "import sys;print(sys.stdout.encoding)"
utf-8
```

I thought you might have some insights (in particular, knowing what encoding you're using might help), but in the meantime, I can muddle along by working on a separate test.
Oh! On that note, I see that you rarely have multiple `page.rst`s for a test module. Would you like me to ultimately extend `code/page.rst`, or create a new `.rst` file?

Thanks for your time.

**Comment by blairconrad** [comment]
The lexer progresses. I'm happy about almost everything, except the question of what to do about mixing the RGB colours into the rest.

1. there's an existing bug where if both the foreground and background colours are specified as RGB, the foreground colour is applied to both
2. allowing background colours to be set via RGB but not codes 40–47 seems odd

I think for now, unless you want to expend to actually support background colours, best would just be to ignore all the background colour-setting codes.
I'll keep on with that unless I hear different. In the meantime, here's a sample:

![image](https://user-images.githubusercontent.com/3275797/82625794-13fd3000-9bb4-11ea-8d3d-23c6b4e8c2f3.png)

**Comment by blairconrad** [comment]
Okay, this is just way too much fun. I'm either exceeding expectations or working toward becoming a huge pain! `^[[38;2;xxx` now supported!

![image](https://user-images.githubusercontent.com/3275797/82705378-8e29c500-9c45-11ea-8cce-77bbcfa2d6b5.png)

**Comment by mosra** [comment]
Merged in #151. Sorry for the delays in my replies, too much to do, too little time :sweat_smile:

**Comment by blairconrad** [comment]
Ha. No sorries! I was working from my fork anyhow in the meantime. I was just happy to have had a chance to play.
